### PR TITLE
[Merged by Bors] - feat: port Topology.Gluing

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
---

This triggers a stack overflow in the compiler.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by adomani** [comment]
There are just a small number of sorries left and a couple of uses of the `continuity` tactic that no longer work.

Any help is greatly appreciated, thanks!

**Comment by Ruben-VandeVelde** [comment]
Stopping now; three errors left

File: Mathlib/Topology/Gluing.lean [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+  f_open : ∀ i j, OpenEmbedding (f i j)
+  f_mono := fun i j => (TopCat.mono_iff_injective _).mpr (f_open i j).toEmbedding.inj
+set_option linter.uppercaseLean3 false in
+#align Top.glue_data TopCat.GlueData
+
+namespace GlueData
+
+variable (D : GlueData.{u})
+
+-- mathport name: «expr𝖣»
```
Comment (semorrison) [inline Mathlib/Topology/Gluing.lean:101]:
```suggestion
```

File: Mathlib/Topology/Gluing.lean [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+
+theorem rel_equiv : Equivalence D.Rel :=
+  ⟨fun x => Or.inl (refl x), by
+    rintro a b (⟨⟨⟩⟩ | ⟨x, e₁, e₂⟩)
+    exacts [Or.inl rfl, Or.inr ⟨D.t _ _ x, by simp [e₁, e₂]⟩], by
+    rintro ⟨i, a⟩ ⟨j, b⟩ ⟨k, c⟩ (⟨⟨⟩⟩ | ⟨x, e₁, e₂⟩); exact id
+    rintro (⟨⟨⟩⟩ | ⟨y, e₃, e₄⟩); exact Or.inr ⟨x, e₁, e₂⟩
+    let z := (pullbackIsoProdSubtype (D.f j i) (D.f j k)).inv ⟨⟨_, _⟩, e₂.trans e₃.symm⟩
+    have eq₁ : (D.t j i) ((pullback.fst : _ /-(D.f j k)-/ ⟶ D.V (j, i)) z) = x := by
+      -- Porting note: was `simp`
```
Comment (semorrison) [inline Mathlib/Topology/Gluing.lean:145]:
```suggestion
      -- Porting note: was `simp`
      -- https://github.com/leanprover-community/mathlib4/issues/5026
```

File: Mathlib/Topology/Gluing.lean [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+    (colimit.ι diagram _ ≫ colim.map _ ≫ (colimit.isoColimitCocone _).hom) _ =
+      (colimit.ι diagram _ ≫ colim.map _ ≫ (colimit.isoColimitCocone _).hom) _ :=
+    (congr_arg
+        (colim.map (diagramIsoParallelPair diagram).hom ≫
+          (colimit.isoColimitCocone (Types.coequalizerColimit _ _)).hom)
+        this :
+      _)
+  -- Porting note: was
+  -- simp only [eqToHom_refl, types_comp_apply, colimit.ι_map_assoc,
+  --   diagramIsoParallelPair_hom_app, colimit.isoColimitCocone_ι_hom, types_id_apply] at this
```
Comment (semorrison) [inline Mathlib/Topology/Gluing.lean:201]:
```suggestion
  -- Porting note: was
  -- simp only [eqToHom_refl, types_comp_apply, colimit.ι_map_assoc,
  --   diagramIsoParallelPair_hom_app, colimit.isoColimitCocone_ι_hom, types_id_apply] at this
  -- See https://github.com/leanprover-community/mathlib4/issues/5026
```

File: Mathlib/Topology/Gluing.lean [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+#align Top.glue_data.ι_mono TopCat.GlueData.ι_mono
+
+theorem image_inter (i j : D.J) :
+    Set.range (𝖣.ι i) ∩ Set.range (𝖣.ι j) = Set.range (D.f i j ≫ 𝖣.ι _) := by
+  ext x
+  constructor
+  · rintro ⟨⟨x₁, eq₁⟩, ⟨x₂, eq₂⟩⟩
+    obtain ⟨⟨⟩⟩ | ⟨y, e₁, -⟩ := (D.ι_eq_iff_rel _ _ _ _).mp (eq₁.trans eq₂.symm)
+    · exact ⟨inv (D.f i i) x₁, by
+        -- Porting note: was `simp [eq₁]`
```
Comment (semorrison) [inline Mathlib/Topology/Gluing.lean:266]:
```suggestion
        -- Porting note: was `simp [eq₁]`
        -- See https://github.com/leanprover-community/mathlib4/issues/5026
```

File: Mathlib/Topology/Gluing.lean [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+
+theorem MkCore.t_inv (h : MkCore) (i j : h.J) (x : h.V j i) : h.t i j ((h.t j i) x) = x := by
+  have := h.cocycle j i j x ?_
+  rw [h.t_id] at this
+  --  porting note: I think that this went smoother than in Lean 3:
+  --  * `exact` was `convert` followed by `ext; simp`
+  --  * the two remaining goals merged, as they were all the same.
+  exact Subtype.eq this
+  rw [h.V_id]
+  trivial
```
Comment (semorrison) [inline Mathlib/Topology/Gluing.lean:375]: Can you turn this into a structured proof. i.e. use `\cdot` to demarcate subgoals.

File: Mathlib/Topology/Gluing.lean [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+-/
+-- Porting note: removed `@[nolint has_nonempty_instance]`
+structure MkCore where
+  {J : Type u}
+  U : J → TopCat.{u}
+  V : ∀ i, J → Opens (U i)
+  t : ∀ i j, (Opens.toTopCat _).obj (V i j) ⟶ (Opens.toTopCat _).obj (V j i)
+  V_id : ∀ i, V i i = ⊤
+  t_id : ∀ i, ⇑(t i i) = id
+                              -- porting note: was vv `@coe (V j i) (U j) _ (t i j x) ∈ V j k`
```
Comment (semorrison) [inline Mathlib/Topology/Gluing.lean:358]: Why is this indented so far?

File: Mathlib/Topology/Gluing.lean [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+/-- (Implementation) the restricted transition map to be fed into `glue_data`. -/
+def MkCore.t' (h : MkCore.{u}) (i j k : h.J) :
+    pullback (h.V i j).inclusion (h.V i k).inclusion ⟶
+      pullback (h.V j k).inclusion (h.V j i).inclusion := by
+  refine' (pullbackIsoProdSubtype _ _).hom ≫ ⟨_, _⟩ ≫ (pullbackIsoProdSubtype _ _).inv
+  · intro x
+    refine' ⟨⟨⟨(h.t i j x.1.1).1, _⟩, h.t i j x.1.1⟩, rfl⟩
+    rcases x with ⟨⟨⟨x, hx⟩, ⟨x', hx'⟩⟩, rfl : x = x'⟩
+    exact h.t_inter _ ⟨x, hx⟩ hx'
+  -- Porting note: was `continuity`
```
Comment (semorrison) [inline Mathlib/Topology/Gluing.lean:391]:
```suggestion
  -- Porting note: was `continuity`, see https://github.com/leanprover-community/mathlib4/issues/5030
```

File: Mathlib/Topology/Gluing.lean [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+    exact h.cocycle i j k ⟨x, hx⟩ hx'
+  -- Porting note : was not necessary in mathlib3
+  f_mono i j := (TopCat.mono_iff_injective _).mpr fun x y h => Subtype.ext h
+set_option linter.uppercaseLean3 false in
+#align Top.glue_data.mk' TopCat.GlueData.mk'
+
+variable {α : Type u} [TopologicalSpace α] {J : Type u} (U : J → Opens α)
+
+-- porting note: removed `include U`
+
```
Comment (semorrison) [inline Mathlib/Topology/Gluing.lean:440]:
```suggestion
```

File: Mathlib/Topology/Gluing.lean [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+/-- We may construct a glue data from a family of open sets. -/
+@[simps! toGlueData_J toGlueData_U toGlueData_V toGlueData_t toGlueData_f]
+def ofOpenSubsets : TopCat.GlueData.{u} :=
+  mk'.{u}
+    { J
+      U := fun i => (Opens.toTopCat <| TopCat.of α).obj (U i)
+      V := fun i j => (Opens.map <| Opens.inclusion _).obj (U j)
+      t := fun i j => ⟨fun x => ⟨⟨x.1.1, x.2⟩, x.1.2⟩, by
+        -- Porting note: was `continuity`.
```
Comment (semorrison) [inline Mathlib/Topology/Gluing.lean:449]:
```suggestion
        -- Porting note: was `continuity`, see https://github.com/leanprover-community/mathlib4/issues/5030
```

File: Mathlib/Topology/Gluing.lean [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+-/
+-- Porting note: removed `@[nolint has_nonempty_instance]`
+structure MkCore where
+  {J : Type u}
+  U : J → TopCat.{u}
+  V : ∀ i, J → Opens (U i)
+  t : ∀ i j, (Opens.toTopCat _).obj (V i j) ⟶ (Opens.toTopCat _).obj (V j i)
+  V_id : ∀ i, V i i = ⊤
+  t_id : ∀ i, ⇑(t i i) = id
+  -- porting note: was vv `@coe (V j i) (U j) _ (t i j x) ∈ V j k`
```
Comment (semorrison) [inline Mathlib/Topology/Gluing.lean:360]: What is the `vv` here? I'm still confused.

File: Mathlib/Topology/Gluing.lean [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+  t_inter : ∀ ⦃i j⦄ (k) (x : V i j), ↑x ∈ V i k → (((↑) : (V j i) → (U j)) (t i j x)) ∈ V j k
+  cocycle :
+    ∀ (i j k) (x : V i j) (h : ↑x ∈ V i k),
+      -- porting note: this v underscore was `↑(t i j x)`, but Lean type-mismatched
```
Comment (semorrison) [inline Mathlib/Topology/Gluing.lean:364]: Again here?

Comment (jjaassoonn) [reply Mathlib/Topology/Gluing.lean:364]: hmm, I am not sure @adomani  and @Ruben-VandeVelde.

File: Mathlib/Topology/Gluing.lean (hunk 10 above)
Comment (semorrison) [reply Mathlib/Topology/Gluing.lean:360]: Oh, is it meant to be an arrowhead pointing at something on the next line? That is way too confusing for me.

Comment (semorrison) [reply Mathlib/Topology/Gluing.lean:360]: (That would explain why previously this line had weird indenting?)

Comment (semorrison) [reply Mathlib/Topology/Gluing.lean:360]:
```suggestion
```

File: Mathlib/Topology/Gluing.lean (hunk 11 above)
Comment (semorrison) [reply Mathlib/Topology/Gluing.lean:364]:
```suggestion
      -- porting note: the underscore in the next line was `↑(t i j x)`, but Lean type-mismatched
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**